File: lib/gate.mjs
// wicked-orchestration — gate precedence (load-bearing seam).
//
// Encodes ADR-0003 / BUILD-SPINE §3.3: the phase gate consumes governance's
// conformance `decision` and resolves the phase transition by precedence.
// Orchestration owns the *transition*; governance owns the *verdict*. The
// arbiter is a verdict CONSUMER, never an author.
//
//   | governance decision      | phase resolves to            |
//   |--------------------------|------------------------------|
//   | deny / reject (hard)     | rejected (approved blocked)  |
//   | allow_with_conditions    | approved_with_conditions     |
//   | allow                    | approved                     |
//   | null / undefined / none  | gate_running (NEVER approve) |
//
// Invariant (the falsifier in ADR-0003): a hard deny/reject can NEVER resolve to
// approved, and absence of a verdict NEVER silently approves — it stays gate_running.
// Engagement level (auto vs human-approval) governs the *reaction*, never whether
// the gate fires or its resolution here.

import { randomUUID } from "node:crypto";

import { getPhase } from "./store.mjs";
import { applyEvent } from "./reducer.mjs";

export const PHASE_REJECTED = "rejected";
export const PHASE_APPROVED_WITH_CONDITIONS = "approved_with_conditions";
export const PHASE_APPROVED = "approved";
export const PHASE_GATE_RUNNING = "gate_running";

/**
 * Resolve a phase gate from a governance conformance decision.
 * @param {"deny"|"reject"|"allow_with_conditions"|"allow"|null|undefined} decision
 * @returns {"rejected"|"approved_with_conditions"|"approved"|"gate_running"}
 */
export function resolveGate(decision) {
  switch (decision) {
    // Hard veto: the approved edge is unavailable.
    case "deny":
    case "reject":
      return PHASE_REJECTED;

    // Allowed, but obligations ride onto the phase (carried by the caller).
    case "allow_with_conditions":
      return PHASE_APPROVED_WITH_CONDITIONS;

    // Clean pass.
    case "allow":
      return PHASE_APPROVED;

    // No verdict yet / governance absent / anything unrecognized:
    // the gate stays running. NEVER silent-approve.
    default:
      return PHASE_GATE_RUNNING;
  }
}

/**
 * Apply a phase gate from a governance ConformanceClaim, driving the actual
 * `gate_running -> {rejected|approved_with_conditions|approved}` transition.
 *
 * This is the enforceable half of ADR-0003, and the enforcement is STRUCTURAL, not
 * happy-path. Two things happen, both routed through the single-writer reducer:
 *   1. `resolveGate` selects the *target* status for THIS transition, and the edge is
 *      validated against ALLOWED_TRANSITIONS (a phase not in `gate_running` is rejected
 *      by reducer step 2).
 *   2. The governing verdict is PERSISTED on the phase as `gate_decision`. A
 *      `deny`/`reject` marker makes the reducer refuse EVERY subsequent approving edge
 *      (`approved` / `approved_with_conditions`) with reason `vetoed_by_governance` —
 *      checked before the from-assertion and the transition table. So the falsifier
 *      `reject ⇒ ¬approved` holds by ANY route/race/surface (a raw `applyEvent`, a stale
 *      command, re-ordered delivery), NOT merely because `resolveGate` chose `rejected`
 *      on the happy path. That marker, not the target selection, is the guarantee.
 *
 * `approved_with_conditions` stays DISTINCT from `approved`: the obligations ride on the
 * phase and the event carries `conditions: true`, so a consumer can always tell the two
 * apart — they are never collapsed to a bare `approved`.
 *
 * Precedence (ADR-0003):
 *   - deny / reject          -> rejected                  (approved STRUCTURALLY blocked: gate_decision veto)
 *   - allow_with_conditions  -> approved_with_conditions  (claim.obligations copied onto the phase; distinct state)
 *   - allow                  -> approved
 *   - no claim / no decision -> stays gate_running         (NEVER silent-approve)
 *
 * Out of scope (Batch D): consuming the claim from a bus-delivered
 * `wicked.conformance.recorded` event. Here the claim is handed directly.
 *
 * @param {string} phaseId  the phase currently in `gate_running`
 * @param {{ decision?:string, obligations?:string[], claim_id?:string }|null|undefined} conformanceClaim
 * @param {{ subscriberId?:string, dataDir?:string, eventId?:string }} [opts]
 * @returns {{
 *   resolved: string,            // the target status resolveGate chose
 *   applied: boolean,            // did the projection actually transition?
 *   phase: object|null,          // the phase row after applying (null if missing)
 *   reason?: string,             // why it did not apply (e.g. no claim, illegal edge, vetoed_by_governance)
 *   obligations: string[],       // obligations carried onto the phase (conditions path)
 *   conditions: boolean          // true iff resolved to approved_with_conditions (distinct from bare approved)
 * }}
 */
export function applyGate(phaseId, conformanceClaim, opts = {}) {
  const dataDir = opts.dataDir;
  const subscriberId = opts.subscriberId;

  const decision =
    conformanceClaim && typeof conformanceClaim === "object" ? conformanceClaim.decision : undefined;
  const resolved = resolveGate(decision);

  // No verdict (or unrecognized): the gate stays running. NEVER silent-approve,
  // and NEVER emit a transition. This is the "governance absent" branch of ADR-0003.
  if (resolved === PHASE_GATE_RUNNING) {
    const phase = getPhase(phaseId, { dataDir });
    return {
      resolved,
      applied: false,
      phase,
      reason: conformanceClaim ? "no_decision_in_claim" : "no_claim",
      obligations: [],
    };
  }

  // Conditions ride onto the phase so a downstream hook/skill can enforce them.
  const obligations =
    resolved === PHASE_APPROVED_WITH_CONDITIONS && Array.isArray(conformanceClaim.obligations)
      ? conformanceClaim.obligations.slice()
      : [];

  // Drive the transition through the single-writer reducer AND persist the governing
  // verdict (`gate_decision`) onto the phase. Persisting the verdict is what makes the
  // gate STRUCTURAL rather than happy-path: once `deny`/`reject` is on the phase, the
  // reducer refuses ANY later approving edge (reason `vetoed_by_governance`) — so the
  // falsifier `reject ⇒ ¬approved` holds by any route/race/surface, not just because
  // resolveGate happened to pick `rejected` here. `allow_with_conditions`/`allow` are
  // persisted too, for completeness and so the consumed verdict is auditable on the phase.
  const event = {
    id: opts.eventId || `gate:${phaseId}:${randomUUID()}`,
    phaseId,
    from: PHASE_GATE_RUNNING, // assert the gate fires only from gate_running
    to: resolved,
    obligations,
    // approved_with_conditions stays DISTINCT from approved: a consumer reads
    // `conditions: true` + the obligations[] to tell them apart (no collapse).
    conditions: resolved === PHASE_APPROVED_WITH_CONDITIONS,
    gate_claim_id: (conformanceClaim && conformanceClaim.claim_id) || null,
    gate_decision: decision ?? null, // the hard marker the reducer vetoes on
  };
  const r = applyEvent(event, { subscriberId, dataDir });
  const phase = getPhase(phaseId, { dataDir });

  return {
    resolved,
    applied: r.applied,
    phase,
    reason: r.reason,
    obligations,
    conditions: resolved === PHASE_APPROVED_WITH_CONDITIONS,
  };
}

File: src/gate.rs
//! Gate precedence — `resolveGate` + `applyGate` ported from `gate.mjs` (ADR-0003).
//!
//! Orchestration owns the *transition*; governance owns the *verdict*. The gate is a verdict
//! CONSUMER, never an author. It consumes an wicked-apps-core [`ConformanceClaim`], persists the verdict
//! as the phase's `gate_decision` (the HARD veto marker), and resolves the transition through the
//! single-writer [`crate::reducer`]:
//!
//! | claim decision                | phase resolves to            |
//! |-------------------------------|------------------------------|
//! | `Deny` (hard)                 | `Rejected` (approve blocked) |
//! | `AllowWithConditions`         | `ApprovedWithConditions` (+obligations) |
//! | `Allow`                       | `Approved`                   |
//! | no claim                      | stays `GateRunning` (NEVER silent-approve) |
//!
//! NOTE on the decision vocabulary: the prototype's verdict strings were
//! `deny | reject | allow_with_conditions | allow`. The wicked-apps-core [`Decision`] type — the verified
//! gate input this crate is built against — collapses the hard-veto case to a single `Deny` variant
//! (there is no separate `Reject`). The hard-veto branch therefore keys on `Decision::Deny`; the
//! ADR's `reject ⇒ ¬approved` invariant is enforced on that variant.

use wicked_apps_core::emit::EmitEvent;
use wicked_apps_core::{ConformanceClaim, Decision};

use crate::domain::PhaseStatus;
use crate::reducer::{apply_event, get_phase, ApplyOutcome, Event};

/// The coarse fact this crate emits when a phase transitions (counts / ids only — no payload
/// content beyond what the bus needs to correlate). Distinct from the prototype's per-edge
/// `wicked.phase.*` names; this is the single orchestration-domain transition fact.
pub const EV_PHASE_TRANSITIONED: &str = "wicked.orchestration.phase_transitioned";

/// Resolve the target phase status from a governance decision (port of `resolveGate`).
///
/// `None` (no claim / no decision) ⇒ `GateRunning`: the gate stays running and NEVER silent-approves.
pub fn resolve_gate(decision: Option<&Decision>) -> PhaseStatus {
    match decision {
        // Hard veto: the approved edge is unavailable; the only resolution is Rejected.
        Some(Decision::Deny) => PhaseStatus::Rejected,
        // Allowed, but obligations ride onto the phase (carried below).
        Some(Decision::AllowWithConditions) => PhaseStatus::ApprovedWithConditions,
        // Clean pass.
        Some(Decision::Allow) => PhaseStatus::Approved,
        // No verdict yet / governance absent: the gate stays running. NEVER silent-approve.
        None => PhaseStatus::GateRunning,
    }
}

/// The outcome of [`apply_gate`] (mirrors the prototype's `applyGate` return).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GateOutcome {
    /// The target status `resolve_gate` chose.
    pub resolved: PhaseStatus,
    /// Did the projection actually transition?
    pub applied: bool,
    /// Why it did not apply (e.g. `no_claim`, an illegal edge, or `vetoed_by_governance`).
    pub reason: Option<String>,
    /// Obligations carried onto the phase (the conditions path).
    pub obligations: Vec<String>,
    /// True iff resolved to `ApprovedWithConditions` (DISTINCT from a bare `Approved`).
    pub conditions: bool,
}

/// Apply a phase gate from a governance [`ConformanceClaim`], driving the actual
/// `GateRunning -> {Rejected | ApprovedWithConditions | Approved}` transition through the
/// single-writer reducer.
///
/// The enforcement is STRUCTURAL, not happy-path (ADR-0003). Two things happen, both routed through
/// the reducer: (1) [`resolve_gate`] selects the target and the edge is validated against the state
/// machine; (2) the governing verdict is PERSISTED on the phase as `gate_decision`. A `Deny` marker
/// makes the reducer refuse EVERY later approving edge (`vetoed_by_governance`) — so
/// `reject ⇒ ¬approved` holds by ANY route, not merely because `resolve_gate` chose `Rejected` here.
///
/// `claim = None` (or a claim the gate cannot use) leaves the phase in `GateRunning` and emits
/// nothing — the "governance absent" branch. NEVER silent-approve.
///
/// On a real transition a coarse [`EV_PHASE_TRANSITIONED`] fact is emitted via the shared
/// fire-and-forget seam (counts / ids only).
pub fn apply_gate<S: wicked_apps_core::GraphRead + wicked_apps_core::GraphWrite>(
    store: &mut S,
    phase_id: &str,
    claim: Option<&ConformanceClaim>,
    event_id: &str,
) -> anyhow::Result<GateOutcome> {
    let decision = claim.map(|c| &c.decision);
    let resolved = resolve_gate(decision);

    // No verdict (or unrecognized): the gate stays running. NEVER silent-approve, NEVER transition.
    if resolved == PhaseStatus::GateRunning {
        return Ok(GateOutcome {
            resolved,
            applied: false,
            reason: Some(if claim.is_some() {
                "no_decision_in_claim".to_string()
            } else {
                "no_claim".to_string()
            }),
            obligations: Vec::new(),
            conditions: false,
        });
    }

    // Conditions ride onto the phase so a downstream hook/skill can enforce them.
    let obligations: Vec<String> =
        if resolved == PhaseStatus::ApprovedWithConditions {
            claim.map(|c| c.obligations.clone()).unwrap_or_default()
        } else {
            Vec::new()
        };

    // Drive the transition through the single-writer reducer AND persist the governing verdict
    // (`gate_decision`). Persisting it is what makes the gate STRUCTURAL: once `Deny` is on the
    // phase, the reducer refuses any later approving edge. Assert `from = GateRunning` so the gate
    // fires only from the gate-running state.
    let event = Event {
        id: event_id.to_string(),
        phase_id: phase_id.to_string(),
        to: resolved,
        from: Some(PhaseStatus::GateRunning),
        obligations: Some(obligations.clone()),
        gate_decision: decision.cloned(),
    };
    let ApplyOutcome {
        applied, reason, ..
    } = apply_event(store, &event)?;

    if applied {
        // Coarse derived fact — counts / ids only. Fire-and-forget; a drop is dead-lettered, not
        // silent (the shared seam guarantees this), and never fails the caller.
        let phase_after = get_phase(store, phase_id)?;
        let status_token = phase_after
            .as_ref()
            .map(|p| p.status.as_token())
            .unwrap_or(resolved.as_token());
        let evt = EmitEvent::new(
            EV_PHASE_TRANSITIONED,
            "wicked-orchestration",
            "orchestration.phase",
            serde_json::json!({
                "phase_id": phase_id,
                "to": status_token,
                "claim_id": claim.map(|c| c.claim_id.as_str()),
                "obligation_count": obligations.len(),
            }),
        );
        let _ = wicked_apps_core::emit::emit_event(&evt);
    }

    Ok(GateOutcome {
        resolved,
        applied,
        reason,
        obligations: obligations.clone(),
        conditions: resolved == PhaseStatus::ApprovedWithConditions,
    })
}
